src/main/java/com/comphenix/protocol/concurrent/PacketTypeMultiMap.java

```
@@ -0,0 +1,165 @@
+
+	/**
+	 * Adds a value to the map, associating it with the {@link PacketType}s
+	 * contained in the specified {@link ListeningWhitelist}. If the value is
+	 * already present in the set (as determined by {@code equals}), it will not be
+	 * added again.
+	 *
+	 * @param key   the whitelist containing the packet types to associate the value
+	 *              with
+	 * @param value the value to be added
+	 * @throws NullPointerException if the key or value is null
+	 */
+	public synchronized void put(ListeningWhitelist key, T value) {
+		Objects.requireNonNull(key, "key cannot be null");
+		Objects.requireNonNull(value, "value cannot be null");
+
+		for (PacketType packetType : key.getTypes()) {
+			this.typeMap.computeIfAbsent(packetType, type -> new SortedCopyOnWriteSet<>()).add(value,
+					new PriorityHolder(key));
+		}
+	}
+
+	/**
+	 * Removes a value from the map, disassociating it from the {@link PacketType}s
+	 * contained in the specified {@link ListeningWhitelist}. If the value is not
+	 * present, the map remains unchanged.
+	 *
+	 * @param key   the whitelist containing the packet types to disassociate the
+	 *              value from
+	 * @param value the value to be removed
+	 * @return a list of packet types that got removed because they don't have any
+	 *         associated values anymore
+	 * @throws NullPointerException if the key or value is null
+	 */
+	public synchronized List<PacketType> remove(ListeningWhitelist key, T value) {
+		Objects.requireNonNull(key, "key cannot be null");
+		Objects.requireNonNull(value, "value cannot be null");
+
+		List<PacketType> removedTypes = new ArrayList<>();
+
+		for (PacketType packetType : key.getTypes()) {
+			SortedCopyOnWriteSet<T, PriorityHolder> entrySet = this.typeMap.get(packetType);
+			if (entrySet == null) {
+				continue;
+			}
+
+			// we shouldn't have empty entrySets
+			assert !entrySet.isEmpty();
```

is this guaranteed? would we want to throw an exception if this isn't true?

---

It should be guaranteed because both the `add` and `remove` methods use the same exclusive lock, ensuring that the execution order remains unchanged. The assert is more of sanity check for unit tests to.

However, I suggest removing the `synchronized` keyword from the method signature and instead wrapping the entire method body in a `synchronized` block, specifically for the `typeMap` field. This approach moves the synchronization onto an object that can't get locked by accident by some external source.